### DISCUSSION TITLE
Extending ex13 / divergence theorem / interfacial resistance

### DISCUSSION THREAD

**Discussion by gatling-nrl** [kinnala/scikit-fem#883] [Q&A]
There's been several previous discussions about calculating integrated flux, and in particular that it would be more stable/accurate to  use the divergence theorem to replace the integral over facets with an integral over cells. The previous conversations were in the context of flux in/out of a subdomain. This discussion is a little different (I think) because I'm working with a subset of the exterior boundary facets instead of a subset of the cells.

Is it possible to change ex13 to use this divergence theorem idea when it is calculating port flux? I can't see how to do that. Is ex13 the best way to do ex13?

I'm also interested in how I might introduce an interfacial resistivity between the "port" and the interior. This would cause the value on the facet to be different than the value just inside the domain (a discontinuity at the facet). This is perhaps part of #867, but I'm asking exactly in the context of ex13.

My application has more than 2 ports, but I assume that if I can get this working for the 2 ports in ex13, I can do it for N ports.

**Comment by gdmcbain** [comment]
G'day.  I'm quite unwell at the moment (yep, covid) but I'll see if I can set you in the right direction.  First, can we split this?  It sounds like two separate problems.  I'll treat the divergence theorem here.

> Is it possible to change ex13 to use this divergence theorem idea when it is calculating port flux?

ex13 does use this divergence theorem idea:

https://github.com/kinnala/scikit-fem/blob/4084c41261840e422fb9af750aace7d721e31cbe/docs/examples/ex13.py#L51

for an explanation, see [&lsquo;Three ways to compute multiport inertance&rsquo;](https://journal.austms.org.au/ojs/index.php/ANZIAMJ/article/view/14058); this is the third way.

> Is ex13 the best way to do ex13?

Well, depends on what you mean by best.  The second of the three ways is slightly more accurate but the third is so much simpler that I pretty much always use the third.

> My application has more than 2 ports, but I assume that if I can get this working for the 2 ports in ex13, I can do it for N ports.

Yep, this is covered in the paper.  It does involve solving N &minus; 1 boundary value problems (which reduces to 2 &minus; 1 = 1 in the two-port case of ex13).  Basically charge one port and earth the rest in each run.

> This discussion is a little different (I think) because I'm working with a subset of the exterior boundary facets instead of a subset of the cells.

I don't _think_ this should matter, but I'm a little befogged today.  Could you provide a sketch?  Is it like you've got a plane region with holes in it except that the holes are filled with highly conductive stuff?  What's on the outer boundary?  Insulated or earthed?  Say earthed.  The idea then is to solve a b.v.p. for each hole with the interior set to one and the interior of all the other holes set to zero.  (Here I'm approximating the highly conductive stuff as perfect; is that a reasonable simplification to get you going?)  You'll find that the matrix for all N problems is the same, only the r.h.s. varies.  Combine the solutions using that matrix as a bilinear form, like U.T @ A @ U, where the columns of U are the solutions of the individual BVPs.  The result is the overall conductance matrix between the N ports; the diagonals are the driving-point conductance, the offs minus the transfer-conductances.

**Reply by gatling-nrl** [reply]
Your paper is making a lot more sense to me this time reading it through. I'm stuck at equation (11) though. It looks like you used greens identity (integration by parts), like this:

![image](https://user-images.githubusercontent.com/38136423/159269054-c02b1a03-f42a-45d9-bb29-b4f50f1bded0.png)

I think doing that depends on your \rho being constant? If you kept a variable coefficient (i.e. \frac{1}{\rho} \nabla u )  how would you get from (11) to (12)?

edit: PS, hope you're feeling better!

edit2: maybe I'm not as sure how you got from (11) to (12) as I thought I was....

edit3: actually, its getting from (12) to (13) when the coef (\rho in the paper) is not constant... and I think therefore eq(19) won't apply when the coef is not constant?

**Reply by gdmcbain** [reply]
(&nabla;*v*, *&rho;*&nabla;*u*)<sub>*&Omega;*</sub> = (*v* **n**, *&rho;*&nabla;*u*)<sub>*&Gamma;*</sub> &minus; (*v*, &nabla;&sdot;{*&rho;*&nabla;*u*})<sub>*&Omega;*</sub> 

Edit: Sorry, that was written without looking at the paper.  That's correct in itself but not what's required for inertance since it's the specific volume *&rho;*<sup>&minus;1</sup> that's analogous to conductivity.

(&nabla;*v*, *&rho;*<sup>&minus;1</sup>&nabla;*u*)<sub>*&Omega;*</sub> = (*v* **n**, *&rho;*<sup>&minus;1</sup>&nabla;*u*)<sub>*&Gamma;*</sub> &minus; (*v*, &nabla;&sdot;{*&rho;*<sup>&minus;1</sup>&nabla;*u*})<sub>*&Omega;*</sub>

**Reply by gatling-nrl** [reply]
First of all, how did you get math in markdown?!

**Reply by gdmcbain** [reply]
In (11), what's written there as **U**<sup>(<i>j</i>)</sup> is a flux, so it's the product of a coefficient (here the specific volume *&rho;*<sup>&minus;1</sup>) and a gradient &nabla;*&Pi;*<sup>(<i>j</i>)</sup>.  Just keep that all together, and (12) follows from (11); for the test flux **U**<sup>(<i>i</i>)</sup> though on the left-hand side of (11), just use the gradient, or divide the flux by the coeffcient (multiply by density here).  So:

(&nabla;*v*, *&rho;*<sup>&minus;1</sup>&nabla;*u*)<sub>*&Omega;*</sub> = (*v* **n**, *&rho;*<sup>&minus;1</sup>&nabla;*u*)<sub>*&Gamma;*</sub> &minus; (*v*, &nabla;&sdot;{*&rho;*<sup>&minus;1</sup>&nabla;*u*})<sub>*&Omega;*</sub> 

Then to get to (13) the inner product over the boundary *&Gamma;* needs to be split up.  The idea is that the integrand will vanish except over the port of interest and therefore this can be related to the &lsquo;reciprocal inertance coefficient&rsquo; *s* (more generally a conductance coefficient).  This vanishing is brought about by the choice of test potential *v* and its boundary conditions; i.e. (6) and (7).

Thank you, I am feeling better; maybe still a bit &lsquo;foggy&rsquo;, I'm not sure.

**Reply by gdmcbain** [reply]
> First of all, how did you get math in markdown?!

Markdown can include HTML, including entities like &amp;nabla; and tags like &lt;sup&gt;&minus;1&lt;/sup&gt;.

**Reply by gdmcbain** [reply]
i.e. in (11) and (12), replace **U**<sup>(<i>i</i>)</sup> with &nabla;*v* (and so *&Pi;*<sup>(<i>i</i>)</sup>/*&rho;* with *v*) and **U**<sup>(<i>j</i>)</sup> with *&rho;*<sup>&minus;1</sup>&nabla;*u*.

**Comment by gatling-nrl** [comment]
So I'm still quite struggling with getting accurate flux measurements out of skfem. Context is still Poisson equation with a variable coefficient. I'm using some output from COMSOL solving a toy problem as a reference. I get extremely poor accuracy using what I now call the "naïve" method of first solving for `u` and then integrating `sigma * grad(u)` over a facet subset. Here is how the hunt for a "better" method as gone so far:

I tried to extend your paper @gdmcbain to handle the variable coef, but without any success so far. I also tried to tinker with [ex13.py](https://github.com/kinnala/scikit-fem/blob/master/docs/examples/ex13.py), but got nowhere when the coefficient was non-constant.

Googling harder, I found this four year old post on scicomp which I could have written word-for-word
https://scicomp.stackexchange.com/questions/26774/computing-accurate-fluxes-with-fem
and that put me on the path of looking into "mixed Poisson problems"

That led me right into [ex37.py](https://github.com/kinnala/scikit-fem/blob/master/docs/examples/ex37.py) which seemed promising, but also doesn't include variable coefficients. I did successfully extend the example though, but then discovered that Dirichlet boundary conditions are somehow hardcoded in... I can't figure by what mechanism though. So i kept digging....

From a 2012 book by Dhatt, I found
![image](https://user-images.githubusercontent.com/38136423/162995672-bfbfdd15-f9c9-4869-bb2c-dce774fffe20.png)
Which made me think I could ditch the RT0 element altogether and treat the (x,y) components of the flux (2d problem) separately. Basically example 37, but with three basis sets: u, flux_x, and flux_y. But this didn't work either.

I'm really beating my head against the wall at this point, because this seems like a well-solved, much discussed problem (calculating accurate fluxes) and I just can't get it to work. I also must admit to being very much a "user" of FEM and not well versed on its guts, or some of the seemingly important math concepts (like notation that reads H(div, Omega) or the concept of H-div conforming). I have no idea how I would extend ElementTriRT0 to ElementTriRT1 or why ElementTriRT0 and ElementTriP2 can't work together. At the moment, all these questions make a deep, deep rabbit hole.

What i really need to do first is specify a P0 sigma in the domain, specify a voltage on a subset of the external boundary facets, specify ground on some other disjoint subset of the boundary, and require zero normal flux everywhere else on the external boundary. Then I need an accurate calculation of the total flux between the two "electrodes" by any practical means. The potential solution inside the domain is irrelevant to me, except to sanity check things.

**Reply by gatling-nrl** [reply]
Code extending ex37.py to handle variable coef. 

```python
import numpy as np
import matplotlib.pyplot as plt
import skfem
from skfem.helpers import dot, div

mesh = skfem.MeshTri().refined(4)
element = skfem.ElementTriRT0() * skfem.ElementTriP0()
basis = skfem.CellBasis(mesh, element)
cb0 = basis.with_element(skfem.ElementTriP0())

sigma = cb0.zeros()
sigma[:] = 1
sigma[cb0.get_dofs(elements=lambda x: x[1]>.5)] = 10

@BilinearForm
def bilinf(u1, v1, u2, v2, w):
    return 1/w.sigma*dot(u1, u2) + div(u1) * v2 + div(u2) * v1

A = asm(bilinf, basis, sigma=cb0.interpolate(sigma))

u0_dofs = basis.get_dofs(facets=[lambda x: x[1]<.01, lambda x: x[1]>.99])
u0 = basis.zeros()
u0[basis.get_dofs(facets=lambda x: x[1]>.99)] = 1
u = skfem.solve(*skfem.enforce(A, x=u0, D=u0_dofs))
(u1, u1_basis), (u2, u2_basis) = basis.split(u)

fig, ax = plt.subplots(1, 3, figsize=(16,5))
cb0.plot(sigma, ax=ax[0], cmap='viridis', shading='gouraud')
u2_basis.plot(u2, ax=ax[1], cmap='viridis', shading='gouraud')
u1_basis.plot(u1, ax=ax[2])
```
![image](https://user-images.githubusercontent.com/38136423/163005980-d768b137-e500-4de5-a41d-13b5c9c0e86d.png)

It almost seems to work as expected, but why are the left and right edges held at zero? I also don't know how to integrate the normal component of `u1` along the top edge. Constructing a facet basis with 

```python
fbasis = skfem.FacetBasis(mesh, element, facets=lambda x: x[1]>.99)
```

fails with 

```
ValueError: operand has more dimensions than subscripts given in einstein sum, but no '...' ellipsis provided to broadcast the extra dimensions.
```
**Edit2**: This is already known by @kinnala in #869 

**Edit**: Also, the P0 solution for `u2` looks terrible. I don't have that kind of visual intuition about the feather plot, but are RT0 and P0 good elements for computing accurate fluxes? (I can't compare to COMSOL yet, because I can't integrate the flux.)

**Reply by gatling-nrl** [reply]
Regarding why I'm working so hard at this: In a circle domain with sigma and solutions that look like this

![image](https://user-images.githubusercontent.com/38136423/163007819-1471b826-0d66-40a2-92ad-b7333a2a9a5d.png)

I vary (left) the voltage on the electrode, (center) the contrast of the two conductivity regions, and (right) the location of the division between conductivities. At each of these, I use the "naïve method" to calculate flux from the top electrode and get

![image](https://user-images.githubusercontent.com/38136423/163008124-f3bf4bd7-fbc4-4260-ab9b-92b191b3ce98.png)

which is clearly wrong (very inaccurate).

**Reply by kinnala** [reply]
Did you try the first approach in scicomp?

**Reply by kinnala** [reply]
Can you write down exactly the boundary value problem that you are trying to solve and then the quantity you are trying to find accurately?

**Reply by gatling-nrl** [reply]
![image](https://user-images.githubusercontent.com/38136423/163043599-d34c9117-23b4-4eaa-a995-5df6318b50dc.png)

**Reply by gatling-nrl** [reply]
In the end 3d. But right now 2d is great, and the domain looks something like this, for L=8. I bolded the electrodes, but it is okay for them to be facets, i.e. a subset of the boundary.

![image](https://user-images.githubusercontent.com/38136423/163044194-8f7fd8ac-4534-4af1-ac7b-70779551cf70.png)

**Reply by gatling-nrl** [reply]
> Did you try the first approach in scicomp?

No, I've focused on the 2nd answer (the one by 56th) because that one seemed to be the method in example 37.

**Reply by gatling-nrl** [reply]
In the latex above, I simplified equation (2), and I can get pretty far the way its written.

In the long run though, I eventually need to replace (2) with 

![image](https://user-images.githubusercontent.com/38136423/163045053-270e3891-1a8d-4cb0-a720-dc0bf7887e97.png)

which is an interfacial condition, as being discussed in #867.  

**Edit**: I don't want to confuse things with this right now though. It's much more important to me to get the simplified problem working and deal with the contact impedance later. I just wanted to put the whole problem down altogether in case it forced some kind of significant change now.

**Reply by gdmcbain** [reply]
> Did you try the first approach in scicomp?

This might not be necessary.  It gives the distribution of flux over the field whereas it's only the integral that's required. That's the difference between Hughes's exercise and the second of our &lsquo;Three ways&rsquo;.

> I tried to extend your paper @gdmcbain to handle the variable coef, but without any success so far. I also tried to tinker with [ex13.py](https://github.com/kinnala/scikit-fem/blob/master/docs/examples/ex13.py), but got nowhere when the coefficient was non-constant.

I'm puzzled at this and don't know what you mean. Is it that as the nonuniformity of the conductivity increases the accuracy deteriorates unacceptably? Or haven't you been able to adapt it to a variable coefficient at all?  I do like the mixed approach too, but I wouldn't have thought it the only possible way.

**Reply by gatling-nrl** [reply]
> the second of our ‘Three ways’.

This is the one I worked at. 

> Or haven't you been able to adapt it to a variable coefficient at all?

I haven't been able to adapt it at all. Here's my derivation so far, trying to follow the 2nd way in your paper.

![image](https://user-images.githubusercontent.com/38136423/163054972-ad97fd94-40d7-4bd8-9632-c770e16ae457.png)

I think my (6) here is your (12). Apologies about the change in notation... I'm trying to stay consistent with my prior work. At this point I have to admit I don't know what the `i` and `j` solutions really are. I tried to roll with it a bit farther

![image](https://user-images.githubusercontent.com/38136423/163055286-14c4d2cb-b6ef-4344-a0fb-39a320e1e744.png)

But now the confusion over `i` and `j` is really showing. I know that in the `j` solution, `U_l=0` when `l!=j` and that seems like it would be useful to kill one of those terms in my (8), but the `i` and `j` are "backwards" from what I need for that to work.

I also assume highly conductive electrodes, so `u` will be constant in the integrand over an electrode. It seems like that would help too.

**Reply by gdmcbain** [reply]
> At this point I have to admit I don't know what the i and j solutions really are. 

The answer is in your (4): the *j*-th solution is *u<sub>j</sub>* &lsquo;when a potential is applied only to *E<sub>j</sub>*&rsquo;.

No worries about changing notation; I don't make any special claims for mine and yours is clear enough.

**Reply by gatling-nrl** [reply]
>  the j-th solution is uj ‘when a potential is applied only to Ej’.

so in my (6) that's the inner product of the (gradient of the) solution when driving the i-th electrode with the (gradient of the) solution when driving the j-th electrode?

So then later, in (8) I would rather break it up on the i-th electrode.. i.e.

![image](https://user-images.githubusercontent.com/38136423/163058158-afbe3d0a-9e96-4a27-8def-19938cf94a0c.png)

and, somewhat excitedly...

![image](https://user-images.githubusercontent.com/38136423/163059529-26611162-a8fe-41fe-9de1-27ba692dfd63.png)

Now if I use my (5) I can write

![image](https://user-images.githubusercontent.com/38136423/163059877-68ad4644-e3e9-413d-b846-034e9f6c89ab.png)

And my (11) is almost identical to your (13). Except my sigma is "back under" the integral. And it can't possibly be that straightforward can it?

In any case, my (10) is just an `skfem.Functional` over two solutions, so I'm going to try that right now.

**Reply by gdmcbain** [reply]
Ah, I see.  Your (2) is different to what's proposed in the &lsquo;Three ways&rsquo;.  Well, it's sort of true.  There is a boundary value problem like (1) &ndash; (3) but then how we approach it is that the single solution *u* of that is decomposed into a linear combination of fundamental solutions like *u<sub>j</sub>* which is the solution of the problem like (1) &ndash; (3) except that (2) is changed so that *U<sub>i</sub>* = &delta;<sub><i>ij</i></sub>; this is (6) of &lsquo;Three ways&rsquo;.

**Reply by gatling-nrl** [reply]
```python

mesh = skfem.io.meshio.from_file(str(mesh_file), None)

electrode_1 = 'b27'
electrode_2 = 'b10'

element = skfem.ElementTriP2()
cb2 = skfem.CellBasis(mesh, element)
cb0 = cb2.with_element(skfem.ElementTriP0())

sigma = cb0.zeros()
sigma[:] = 1
sigma[cb0.get_dofs(elements='top')] = 10

@skfem.BilinearForm
def laplace(u, v, w):
    return dot(w.sigma * grad(u), grad(v))

A = laplace.assemble(cb2, sigma=cb0.interpolate(sigma))

u0_dofs = cb2.get_dofs(facets=[electrode_1, electrode_2])
u0 = cb2.zeros()
u0[cb2.get_dofs(facets=electrode_1)] = 1
u1 = skfem.solve(*skfem.enforce(A, x=u0, D=u0_dofs))

u0_dofs = cb2.get_dofs(facets=[electrode_1, electrode_2])
u0 = cb2.zeros()
u0[cb2.get_dofs(facets=electrode_2)] = 1
u2 = skfem.solve(*skfem.enforce(A, x=u0, D=u0_dofs))

@skfem.Functional
def measure_flux(w):
    return w.sigma * dot(grad(w.u1), grad(w.u2))

flux = measure_flux.assemble(cb2, u1=u1, u2=u2, sigma=cb0.interpolate(sigma))
print(f'{-flux:.4f}  ({(-flux - df1.current.iloc[-1])/df1.current.iloc[-1] * 100:.2f}% from COMSOL)')


fig, ax = plt.subplots(1, 3, figsize=(12,4))
cb0.plot(sigma, cmap='viridis', shading='gouraud', ax=ax[0])
mesh.draw(ax=ax[0])
cb2.plot(u1, cmap='viridis', ax=ax[1])
cb2.plot(u2, cmap='viridis', ax=ax[2])
```

![image](https://user-images.githubusercontent.com/38136423/163062013-2c62f303-5bf0-4f0d-abeb-e17e801f73cb.png)
```0.7997  (-1.26% from COMSOL)```

The error is _much_ better, but I'd like to drive it even smaller, maybe below .1%...


**Edit**: even the "third way" seems to be happy now...
```python
print(-flux)

flux_way3 = u1 @ A @ u1
print(flux_way3)

flux_way3 = u2 @ A @ u2
print(flux_way3)
```
```
0.7996869483005843
0.7996869483006155
0.7996869483036395
```

**Reply by gatling-nrl** [reply]
Over a range of contrasts in sigma:

![image](https://user-images.githubusercontent.com/38136423/163062717-577dcc8a-1b7e-4b71-ba1b-df1cb94a6237.png)

**Reply by gatling-nrl** [reply]
> except that (2) is changed so that Ui = δij; this is (6) of ‘Three ways’.

Well, in the long run, I really want for (2)

![](https://user-images.githubusercontent.com/38136423/163045053-270e3891-1a8d-4cb0-a720-dc0bf7887e97.png)

But I'm not ready for that complexity yet...

**Reply by gdmcbain** [reply]
> Over a range of contrasts in sigma:

O. K , that's looking pretty good. Thank you.

Why is COMSOL the datum? Do we know what it's doing? Possibly better to assess the mesh convergence of each, as in the &lsquo;Three ways&rsquo;, or cook up a problem with a known exact solution.

**Reply by gatling-nrl** [reply]
> Why is COMSOL the datum? Do we know what it's doing?

Good questions, and no, I don't know much about what its doing. I have its mesh, and that it's using P2 elements, but that's about it. It is fairly opaque about how it computes flux. That being said, it has been a stable sanity check while I've been learning.

I can do that mesh convergence. I'll post some more data shortly.

**Reply by gatling-nrl** [reply]
![image](https://user-images.githubusercontent.com/38136423/163073882-d4a02d77-d321-4387-aeca-7c16aeef10d2.png)

**Reply by gatling-nrl** [reply]
![image](https://user-images.githubusercontent.com/38136423/163074978-7e15d9e1-acc8-4c1d-9c92-2176508ef9ae.png)

Edit:
The small mesh size gets quite sluggish (mostly due to gmsh i think). Here are the extremes used in the above convergence plots:

![image](https://user-images.githubusercontent.com/38136423/163075325-87fb792f-4d02-4823-93ed-73b3a3f478c6.png)

Tomorrow I will try to get some similar data out of COMSOL, because I think it is converging much faster. Obviously I would prefer to work with as coarse a mesh as possible...

**Reply by gdmcbain** [reply]
> The small mesh size gets quite sluggish (mostly due to gmsh i think)

Could be, but also it will get sluggish due to the use of the direct solver.

> Obviously I would prefer to work with as coarse a mesh as possible...

Yes, of course.  The mesh-convergence study here is mainly to get an idea of whether everything is working correctly.  We can already see that the method not marked &lsquo;naive&rsquo; is doing much bettter on the coarser meshes than the one that is.  Is that one the second or third way?  Anyway, yes, when I first started using the second it was very much because itt gave reasonable engineering results (&plusmn;1&ndash;10%) on reasonable engineering (i.e. coarse) meshes.

If you do find that you want to get quicker results on finer meshes, one idea is switching from a direct to an iterative solver; e.g. algebraic multigrid #264.  Fortunately this is really easy in scikit-fem #273.

But once you've got decent results which seem to be converging, as they do, you can try extrapolating from the last few to get the zero-mesh size limit.  A nice tool for this is NASA's `VERIFY`, wrapped for Python as [convergence](https://pypi.org/project/convergence/).  Then instead of the flux, you can plot the discrepancy and plot that and the mesh parameter logarithmically, as in figure 3 of the &lsquo;Three ways&rsquo;.  The solid orange, green, and red curves above are already looking pretty good; I can't tell from the graph whether they're &plusmn;1% or &plusmn;0.1%.

**Reply by gdmcbain** [reply]
> Well, in the long run, I really want for (2)

For this, i.e. changing from Dirichlet to Robin conditions on the electrodes, what I'd think about doing is first solve the complete set of *n* Dirichlet problems to construct the conductance matrix for the *n* electrodes and then think of that as a purely resistive circuit and then add another resistor branch out from the Dirichlet electrode to the real terminal.  That is, I'd work at the lumped-element circuit level rather than the finite-element level.

**Reply by gatling-nrl** [reply]
> Is that one the second or third way?

naïve in that plot is

```python
    @skfem.Functional
    def measure_flux_naive(w):
        return dot(w.sigma * grad(w.u), w.n)

    measure_flux_naive.assemble(fb2, u=fb2.interpolate(u1), sigma=fb0.interpolate(sigma))
```

non-naïve is

```python
    @skfem.Functional
    def measure_flux(w):
        return w.sigma * dot(grad(w.u1), grad(w.u2))

   measure_flux.assemble(cb2, u1=u1, u2=u2, sigma=cb0.interpolate(sigma))
```
